### PR TITLE
Single message notification could use contact avatar

### PR DESCRIPTION
**Pull request by Mvishnu9**
This is a possible fix for #2314 (it should be compatible with #2315 as well)

### CODE REVIEW

**Comment by Mvishnu9** [comment]
I have modified/adapted tests to test some new functionality and to add "when(controller.getContext()).thenReturn(RuntimeEnvironment.application)" wherever that is needed but is missing (it is already there in several cases). But 
calls like "context.getString(R.string.notification_action_reply)" are still resulting in the following type of error (even though "when(controller.getContext()).thenReturn(RuntimeEnvironment.application)" and "context=controller.getContext()" are present earlier in the execution sequence):
android.content.res.Resources$NotFoundException: Unable to find resource ID #0x7f0702c0 in packages [android, org.robolectric.default]

And calls like "context.getResources().getQuantityString(R.plurals.notification_new_messages_title, newMessagesCount, newMessagesCount)" are still resulting in NPE (again, "when(controller.getContext()).thenReturn(RuntimeEnvironment.application)" and "context=controller.getContext()" are present earlier in the execution sequence). 

Please help.

File: k9mail/src/main/java/com/fsck/k9/notification/BaseNotifications.java [hunk 1]
```diff
@@ -50,6 +67,29 @@ protected BaseNotifications(NotificationController controller, NotificationActio
...
+        canvas.drawOval(rectF, paint);
+
+        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
+        canvas.drawBitmap(bitmap, rect, rect, paint);
+
+        bitmap.recycle();
+
+        return output;
+    }
+
```
Comment (Mvishnu9) [inline k9mail/src/main/java/com/fsck/k9/notification/BaseNotifications.java]: The above method is from https://goo.gl/yu6n2k

File: k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java [hunk 2]
```diff
@@ -2,15 +2,26 @@
...
 
 import android.content.Context;
+import android.graphics.Bitmap;
+import android.graphics.Canvas;
+import android.graphics.Color;
+import android.graphics.Paint;
+import android.graphics.PorterDuff;
+import android.graphics.PorterDuffXfermode;
+import android.graphics.Rect;
+import android.graphics.RectF;
```
Comment (Mvishnu9) [inline k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java]: These imports are no longer needed here (remnant of an earlier implementation which I had tried).

**Comment by rhari991** [comment]
The Android components like context and resources cannot be accessed in tests as they run directly on the JVM. You will need to mock them - have a look at the other unit tests to see how this is done.

**Comment by Mvishnu9** [comment]
Thanks but I think I have already done that (or verified that it is being done) wherever it is needed in these cases (I have checked again several times). The following is the stack trace of the first exception:

android.content.res.Resources$NotFoundException: Unable to find resource ID #0x7f0702be in packages [android, org.robolectric.default]

	at org.robolectric.shadows.ShadowAssetManager.getResName(ShadowAssetManager.java:868)
	at org.robolectric.shadows.ShadowAssetManager.resolveResourceValue(ShadowAssetManager.java:676)
	at org.robolectric.shadows.ShadowAssetManager.resolve(ShadowAssetManager.java:636)
	at org.robolectric.shadows.ShadowAssetManager.getAndResolve(ShadowAssetManager.java:630)
	at org.robolectric.shadows.ShadowAssetManager.getResourceText(ShadowAssetManager.java:226)
	at android.content.res.AssetManager.getResourceText(AssetManager.java)
	at android.content.res.Resources.getText(Resources.java:295)
	at android.content.res.Resources.getString(Resources.java:385)
	at android.content.Context.getString(Context.java:377)
	at com.fsck.k9.notification.WearNotifications.addMarkAllAsReadAction(WearNotifications.java:59)
	at com.fsck.k9.notification.WearNotifications.addSummaryActions(WearNotifications.java:43)
	at com.fsck.k9.notification.WearNotificationsTest.testAddSummaryActionsWithDeleteAllActionEnabled(WearNotificationsTest.java:177)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:538)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:339)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:259)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:41)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:199)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:262)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)

**Comment by philipwhiuk** [comment]
That's because when your IDE runs that test the path is wrong. The actual reasons the test fail is because:

```
controller.getContactPictureLoader()
```

returns null

**Comment by Mvishnu9** [comment]
Yes, I had fixed that particular issue (of controller.getContactPictureLoader() returning null) in some tests like BaseNotificationsTest before submitting the PR. I later saw that it is occurring in other tests also. I am now replicating the approach I used in BaseNotificationsTest to other tests. After I did that WearNotificationsTest, the NPE stopped occurring and now I am getting the exception mentioned earlier (android.content.res.Resources$NotFoundException: Unable to find resource ID #0x7f0702be in packages [android, org.robolectric.default] ). It may have something to do with what you have said about the path being wrong when the IDE runs the test. I will look into it. Thanks.

**Comment by Mvishnu9** [comment]
"Regardless I think we are likely to go with the other PR at this point."
@philipwhiuk Just saw the above edit. I will stop working on it further.

**Comment by Valodim** [comment]
The way of drawing a bitmap to an ImageView seems like a hack. Considering that the drawing happens in a background thread I'm not sure what the best way to do this properly is, but it should probably some method that loads synchronously in `Contacts`. Sort of related is also #2315 that moved us to Glide loading for all things contact picture.

**Comment by Mvishnu9** [comment]
@Valodim Yes, the fact that currently the drawing happens in a background thread means that the bitmap does not get drawn sometimes (I observed this during my testing). I thought #2315 took care of this issue. I tried to make my implementation consistent with the way contact picture loading is done in the rest of the application, so that the K9.showContactPicture() setting is taken into account and the bitmap cache gets used.

File: k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java [hunk 3]
```diff
@@ -161,4 +180,8 @@ NotificationManagerCompat getNotificationManager() {
     NotificationCompat.Builder createNotificationBuilder() {
         return new NotificationCompat.Builder(context);
     }
+
+    protected ImageView getNewImageView() {
```
Comment (Mvishnu9) [inline k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java]: Did it this way to allow ImageView to be mocked in test cases.

**Comment by cketti** [comment]
Thanks for working on implementing this feature :+1:

See https://github.com/k9mail/k-9/pull/2430#issuecomment-572212868 for how we plan to proceed.

**Pull request closed without merging**